src/libstore/nar-info-disk-cache.cc

```diff
@@ -189,23 +216,42 @@ class NarInfoDiskCacheImpl : public NarInfoDiskCache
             // the cache for this URI in the meantime.
             auto cache(queryCacheRaw(*state, uri));
 
-            if (cache)
+            if (cache && cacheIsUpToDate(*cache))
                 return;
 
-            state->insertCache.use()(uri)(time(0))(storeDir)(wantMassQuery)(priority).exec();
-            assert(sqlite3_changes(state->db) == 1);
-            state->caches[uri] = Cache{(int) sqlite3_last_insert_rowid(state->db), storeDir, wantMassQuery, priority};
+            Cache ret {
+                // id: to be set in the conditional below
+                .id = -1,
+                // lastCacheInfoFetch: always the current time, because this is
+                // only called when a cached Cache wasn't available, so the parameters
+                // are always fresh.
```

Comment:
I would prefer for this code not to make an assumption, but rather refactor these methods so that `NarInfoDiskCache` only exposes a single method that's an upsert with a lazy default. I've opted for a less invasive fix for now.